cmd/outline-ss-server/metrics.go

```diff
@@ -49,49 +58,158 @@ type outlineMetrics struct {
 var _ service.TCPMetrics = (*outlineMetrics)(nil)
 var _ service.UDPMetrics = (*outlineMetrics)(nil)
 
+// Converts a [net.Addr] to an [IPKey].
+func toIPKey(addr net.Addr, accessKey string) (*IPKey, error) {
+	hostname, _, err := net.SplitHostPort(addr.String())
+	if err != nil {
+		return nil, fmt.Errorf("failed to create IPKey: %w", err)
+	}
+	ip, err := netip.ParseAddr(hostname)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create IPKey: %w", err)
+	}
+	return &IPKey{ip, accessKey}, nil
+}
+
+// Represents the clients that are or have been active recently. They stick
+// around until they are inactive, or get reported to Prometheus, whichever
+// comes last.
+type activeClient struct {
+	info      ipinfo.IPInfo
+	connCount int // The active connection count.
+	startTime time.Time
+}
+
+type IPKey struct {
+	ip        netip.Addr
+	accessKey string
+}
+
+type tunnelTimeCollector struct {
+	ip2info       ipinfo.IPInfoMap
+	mu            sync.Mutex // Protects the activeClients map.
+	activeClients map[IPKey]*activeClient
+
+	tunnelTimePerKey      *prometheus.CounterVec
+	tunnelTimePerLocation *prometheus.CounterVec
+}
+
+func (c *tunnelTimeCollector) Describe(ch chan<- *prometheus.Desc) {
+	c.tunnelTimePerKey.Describe(ch)
+	c.tunnelTimePerLocation.Describe(ch)
+}
+
+func (c *tunnelTimeCollector) Collect(ch chan<- prometheus.Metric) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	tNow := now()
+	for ipKey, client := range c.activeClients {
+		c.reportTunnelTime(ipKey, client, tNow)
+	}
+	c.tunnelTimePerKey.Collect(ch)
+	c.tunnelTimePerLocation.Collect(ch)
+}
+
+// Calculates and reports the tunnel time for a given active client.
+func (c *tunnelTimeCollector) reportTunnelTime(ipKey IPKey, client *activeClient, tNow time.Time) {
+	tunnelTime := tNow.Sub(client.startTime)
+	logger.Debugf("Reporting tunnel time for key `%v`, duration: %v", ipKey.accessKey, tunnelTime)
+	c.tunnelTimePerKey.WithLabelValues(ipKey.accessKey).Add(tunnelTime.Seconds())
+	c.tunnelTimePerLocation.WithLabelValues(client.info.CountryCode.String(), asnLabel(client.info.ASN)).Add(tunnelTime.Seconds())
+	// Reset the start time now that the tunnel time has been reported.
+	client.startTime = tNow
+}
+
+// Registers a new active connection for a client [net.Addr] and access key.
+func (c *tunnelTimeCollector) startConnection(ipKey IPKey) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	client, exists := c.activeClients[ipKey]
+	if !exists {
+		clientInfo, _ := ipinfo.GetIPInfoFromIP(c.ip2info, net.IP(ipKey.ip.AsSlice()))
+		client = &activeClient{info: clientInfo, startTime: now()}
+	}
+	client.connCount++
+	c.activeClients[ipKey] = client
```

Comment:
We can avoid the extra lookup in the happy case:
```suggestion
		c.activeClients[ipKey] = client
	}
	client.connCount++
```

Reply:
You'ree right. This was from before I moved to pointers in the map. Thanks.